src/net/dns.c

```diff
@@ -0,0 +1,151 @@
+#include "dns.h"
+#include <core/debug.h>
+#include <net/ethernet.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static uint16_t dns_id = 1;
+
+void dns_request(net_interface_t* interface, char* domain)
+{
+  // TODO: fixme, we don't have the concept of port/socket yet.
+  uint16_t src_port = 42796;
+
+  dns_header_t dns_header = {
+    .id = HTONS(dns_id),
+    .qdcount = HTONS(0x0001),
+    .flags = HTONS(0x0100), // recursion requested
+  };
+
+  // Transforms a domain name into a DNS query name, e.g.:
+  // foo.example.com -> 3foo7example3com
+  uint32_t domain_len = strlen(domain);
+  uint32_t data_len = (1            // first count
+                       + domain_len // dots will be replaced with counts
+                       + 1          // END
+                       + 2          // Type
+                       + 2          // Class
+                       ) *
+                      sizeof(uint8_t);
+  uint8_t* data = malloc(data_len);
+
+  char* _domain = calloc(1, (domain_len + 1) * sizeof(char));
+  strcat(_domain, domain);
+  strcat(_domain, ".");
+
+  uint32_t j = 0, pos = 0;
+  for (uint32_t i = 0; i < domain_len + 1; i++) {
+    if (_domain[i] == '.') {
+      // add the count first
+      data[j++] = i - pos;
+
+      // copy part
+      while (pos < i) {
+        data[j++] = _domain[pos++];
+      }
+
+      pos++;
+    }
+  }
+  // END
+  data[j++] = 0;
+  // Type
+  data[j++] = 0;
+  data[j++] = 0x01;
```

Comment:
Maybe we should have a struct that would containth qname and then these attributes